[s3_storage] Improvements

hmrodrigues opened the pull request:
I've added some improvements to s3_storage module:

- Added pylint based on OCA rules
- Only write/read/delete from s3 if all configs are set
- If the configuration isn't correct, it defaults to Odoo default behaviour (via super)
- Replaced code from v10 to v9 (v9 lacks garbage collector)

If you want, I can stash the commits.

Best Regards,
Hugo Rodrigues.

hmrodrigues committed:
[REF] Added coauthor and fixed module name on license notice

hmrodrigues committed:
[ADD] pylintrc

hmrodrigues committed:
[REF] Added a more explicit message

hmrodrigues committed:
[IMP] Check if S3 is configured

Override _storage method to ensure that s3 params are correct. If not,
it defaults to file

hmrodrigues committed:
[IMP] Only run S3 code when location is s3 storage

If the storage location isn't set to s3, or the configuration isn't
correct, it calls the super in order to store/read/delete attachements
on the correct location

hmrodrigues committed:
[REF] Replaced v10 methods

It was using methods that only works on v10 and up

hmrodrigues committed:
[REF] Refactor pylint

andreparames commented:
Thanks a lot for your contribution!

I'd prefer if you didn't stash, makes it easier to re-enable GC for the v10 port.

Out of curiosity, will you be using this in production?

hmrodrigues committed:
[FIX] Added missing decorator and parameter

andreparames commented on s3_storage/models/s3storage.py:
```diff
@@ -50,64 +63,52 @@ def _get_storage_client(self):
-        kept = set()
+        self.env.cr.execute("""
+        SELECT COUNT(*) FROM ir_attachment WHERE store_fname = %s""", (fname,))
+        fnames = [r[0] for r in self.env.cr.fetchall()]
```
Won't this result in `fnames` containing just the number of attachments? Also, since we're now removing one file at a time, we should probably use `client.remove_object()` instead, which takes only one filename.

hmrodrigues replied:
True. On the delete method I backported to the v9 method, i'm still testing this

hmrodrigues replied:
fixed on f98628b

andreparames commented on s3_storage/models/s3storage.py:
```diff
@@ -50,64 +63,52 @@ def _get_storage_client(self):
+                            error.error_code,
+                            error.error_message)
+        except ResponseError as errn:
+            LOGGER.info("_file_gc (s3)", exc_info=True)
```
Should be changed to `_file_delete`.

hmrodrigues replied:
Changed on #6c28db81de75a3b191a0eaabdafc246cdec7a1c0

hmrodrigues replied:
fixed on 6c28db8

hmrodrigues committed:
[FIX] Added missing parameter

hmrodrigues commented:
Yes, the idea is to port to v11 and then use it in production.

I'm going to create a method to move all ir.attachments to s3

hmrodrigues committed:
[FIX] Type on self.env

hmrodrigues committed:
[REF] Renamed v9 methods refereces

hmrodrigues committed:
[IMP]

Allows unsecure connections
Fixes missing parameter no _get_path
only deletes on file

hmrodrigues committed:
[FIX] secure flag

hmrodrigues committed:
[IMP] _file_delete is working

andreparames commented on s3_storage/models/s3storage.py:
```diff
@@ -99,17 +102,14 @@ def _file_delete(self, fname):
         except ResponseError as errn:
             LOGGER.info("_file_gc (s3)", exc_info=True)
-        LOGGER.info("filestore  %d removed", len(fnames))
+        LOGGER.info("filestore  %d removed", fname)
```
`%d` → `%s`

hmrodrigues replied:
fixed on  f98628b

andreparames commented on s3_storage/models/s3storage.py:
```diff
@@ -1,113 +1,110 @@
     @api.model
     def _get_storage_client(self):
         config = self.env['ir.config_parameter'].sudo()
+        secure = config.get_param('s3_attachment_secure', True)
```
Since you're adding a new key, it'd be nice to update the README.

hmrodrigues replied:
Sorry I forgot. added on 7374dab

hmrodrigues committed:
[ADD] Added s3_attachment_secure config to README

andreparames commented:
LGTM :+1: Are you ready to merge?

hmrodrigues commented:
@andreparames not yet, the _file_read method is not working. it's giving an error about date format

andreparames commented:
Date format? Curious. Feel free to post it if you want a second pair of eyes on it.

hmrodrigues commented:
When we check for the file size, minio api gives the following error
```
  File "/home/hmr/Programming/odoo/odoo/openerp/addons/base/ir/ir_attachment.py", line 142, in _data_get
    result[attach.id] = self._file_read(cr, uid, attach.store_fname, bin_size)
  File "/home/hmr/Programming/odoo/odoo/openerp/api.py", line 250, in wrapper
    return old_api(self, *args, **kwargs)
  File "/home/hmr/Programming/odoo/odoo/openerp/api.py", line 354, in old_api
    result = method(recs, *args, **kwargs)
  File "/home/hmr/Programming/odoo/odoo/myaddons/s3_storage/models/s3storage.py", line 76, in _file_read
    info = client.stat_object(bucket, fname)
  File "/usr/lib/python2.7/site-packages/minio/api.py", line 962, in stat_object
    last_modified = strptime(last_modified, http_time_format)
  File "/usr/lib/python2.7/_strptime.py", line 478, in _strptime_time
    return _strptime(data_string, format)[0]
  File "/usr/lib/python2.7/_strptime.py", line 332, in _strptime
    (data_string, format))
ValueError: time data 'Mon, 26 Feb 2018 15:56:30 GMT' does not match format '%a, %d %b %Y %H:%M:%S GMT'
```
The format is hardocoded on [minio api](https://github.com/minio/minio-py/blob/master/minio/api.py#L961)

I'm actually using minio locally

andreparames commented:
Looks like a bug in Minio when running in non-English locales. The `%a` and `%b` directives are locale-specific, so won't recognize "Mon" if you're running in a different locale.

Try adding the following before the call to minio, just for confirming:

```
import locale
locale.setlocale(locale.LC_TIME, "en")
```

hmrodrigues commented:
Hmmm it fixes the error but it will depend on the system where Odoo is running, in my case I don't have the en locale so I had to use en_GB.UTF-8. How about to use a system parameter so users can set an available locale? If the parameter isn't set, it defaults to en

andreparames commented:
The code was just to confirm the problem; it's not actually a good solution. The bug should be fixed by the Minio library itself, which should not blindly use strptime() to parse dates in English. Do you mind opening an issue in [their tracker](https://github.com/minio/minio-py/issues)?

hmrodrigues commented:
True story.

I think that to be production ready, we need to check if the attachment is on s3 or maybe upload all attachments to s3. What do you think is the best solutions?

andreparames commented:
I think the best is to check, and also to have a manual way to migrate. I think Odoo already has something like that (for filesystem vs database storage).

andreparames commented:
Looks like I was misremembering, there are only a few XML-RPC scripts to do that: https://www.odoo.com/forum/help-1/question/how-to-convert-from-filestore-to-databasestore-711

I don't know, see what's best for your use case :)

hmrodrigues commented:
They have a method to [do that](https://github.com/odoo/odoo/blob/9.0/openerp/addons/base/ir/ir_attachment.py#L59) (migrate from db to fs and vice-versa), but we can't extend a method to just customize the domain and add a new storage location. The upstream method will work for existing database store, since **s3** and **file** locations work the same way.

A possible solution is to create a new method to force storage on the filestore and then upload to s3.

hmrodrigues committed:
[IMP] Store files inside database path.

This allows to have a bucket for multi databases

hmrodrigues committed:
[FIX] Using relative path

hmrodrigues committed:
[ADD] method to move all attachments to s3

hmrodrigues committed:
[FIX] move_to_s3 method

Fixed the method to move all attachments to S3. It can be called from
xmlrpc or S3. Since we force the storage to be file before moving, the
permissions are checked by the upstream method. To delete the files
localy, the user must pass the argument delete_fs as True

hmrodrigues commented:
@andreparames I've added a method to copy/move attachments to S3.

I'm ready to merge.

andreparames commented:
Thanks! It's merged.
